database/models.go

```diff
@@ -71,17 +135,78 @@ type Vulnerability struct {
 	Severity    Severity
 
 	Metadata MetadataMap
+}
+
+// VulnerabilityWithAffected is an vulnerability with all known affected
+// features.
+type VulnerabilityWithAffected struct {
+	Vulnerability
 
-	FixedIn                        []FeatureVersion
-	LayersIntroducingVulnerability []Layer
+	Affected []AffectedFeature
+}
+
+// PagedVulnerableAncestries is a vulnerability with a page of affected
+// ancestries each with a special index attached for streaming purpose. The
+// current page number and next page number are for navigate.
+type PagedVulnerableAncestries struct {
+	Vulnerability
 
-	// For output purposes. Only make sense when the vulnerability
-	// is already about a specific Feature/FeatureVersion.
-	FixedBy string `json:",omitempty"`
+	// Affected is a map of special indexes to Ancestries, which the pair
+	// should be unique in a stream. Every indexes in the map should be larger
+	// than previous page.
+	Affected map[int]string
+
+	Limit   int
+	Current PageNumber
+	Next    PageNumber
 }
 
+// Notification is a hook used to notify and give token to user to retrieve the
+// detailed notification e.g. vulnerability changes.
+type Notification struct {
+	Name string
+
+	Created  time.Time
+	Notified time.Time
+	Deleted  time.Time
+}
+
+// VulnerabilityNotification is a notification for vulnerability changes.
+type VulnerabilityNotification struct {
+	Notification
+
+	Old *Vulnerability
+	New *Vulnerability
+}
+
+// PageNumber is used to do pagination.
+type PageNumber struct {
+	// StartID is an implementation detail for paginating by an ID required to
+	// be unique to every ancestry and always increasing.
+	StartID int
```

Comment:
ok, so in that case the implementation should initialize an encrypted version of
 VulnerabilityNotificationFirstPage and VulnerabilityNotificationLastPage 
when it's registered.